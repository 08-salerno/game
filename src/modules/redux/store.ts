import { configureStore } from '@reduxjs/toolkit';
import createSagaMiddleware from 'redux-saga';
import userReducer from './slices/userSlice';
import rootSaga from './sagas/root.saga';

const sagaMiddleware = createSagaMiddleware();

const store = configureStore({
  reducer: {
    user: userReducer,
  },
  middleware: [sagaMiddleware],
});

sagaMiddleware.run(rootSaga);

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch

export default store;
