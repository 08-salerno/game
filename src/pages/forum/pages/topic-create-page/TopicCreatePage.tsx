import React from 'react';
import {
  ErrorMessage, Field, FieldProps, Form, Formik, FormikValues,
} from 'formik';
import { object, string, StringSchema } from 'yup';
import { createTopic } from '../../api';
import ForumRoutes from '../../routes';
import { FormikSubmit } from '../../../../modules/utils/formik.utils';
import useAppRouter from '../../../../modules/router/router';
import styled from 'styled-components';

const TopicCreateSchema = object().shape<{ title: StringSchema }>({
  title: string().required('').min(3, 'Минимум 3 символа'),
});

const MaxTitleLength = 300;

interface TopicCreateFormValue extends FormikValues {
  title: string;
}

const TopicCreatePage: React.VFC = () => {
  const router = useAppRouter();

  const handleSubmit: FormikSubmit<TopicCreateFormValue> = (
    values: TopicCreateFormValue,
  ): Promise<void> => createTopic(values.title)
    .then((id: string) => {
      router.go(`${ForumRoutes.HOME}/${id}`);
    })
    .catch(() => {
      // todo [sitnik] уведомить пользоователя об ошибке
    });

  const CreateInput = styled.input`
    font-family: Arial;
    width: 100%;
    border: none;
    border-bottom: 1px solid ${(props): string => props.theme.form.underline};
    color: ${(props): string => props.theme.form.font};
    background-color: ${(props): string => props.theme.form.background};
    font-weight: 500;
    font-size: 14px;
    line-height: 16px;
  `;
  return (
    <Formik
      initialValues={{ title: '' }}
      validationSchema={TopicCreateSchema}
      onSubmit={handleSubmit}
    >
      {({ isSubmitting, isValid, dirty }): React.ReactElement => (
        <Form>
          {/* todo [sitnik] вынести в переиспользуемый компонент*/}
          <Field name="title" maxLength={MaxTitleLength} placeholder="Описание">
            {(props: FieldProps<string, TopicCreateFormValue>): React.ReactElement => (
              <div>
                {/* eslint-disable-next-line react/jsx-props-no-spreading */}
                <textarea {...props.field} />
                <div>
                  {props.field.value.length || 0}/{MaxTitleLength}
                </div>
              </div>
            )}
          </Field>
          <div>
            <ErrorMessage name="title" />
          </div>
          <div>
            <button type="submit" disabled={isSubmitting || !(dirty && isValid)}>
              Создать тему
            </button>
          </div>
        </Form>
      )}
    </Formik>
  );
};

export default TopicCreatePage;
