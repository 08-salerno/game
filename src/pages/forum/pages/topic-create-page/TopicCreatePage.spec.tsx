import React from 'react';
import { render } from 'enzyme';
import { Provider } from 'react-redux';
import TopicCreatePage from './TopicCreatePage';
import store from '../../../../modules/redux/store';

it('Auth', () => {
  expect(render(<Provider store={store}>
                    <TopicCreatePage />
                </Provider>)).toMatchSnapshot();
});
