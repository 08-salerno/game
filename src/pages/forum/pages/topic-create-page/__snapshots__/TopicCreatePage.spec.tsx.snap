// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth 1`] = `
LoadedCheerio {
  "0": Node {
    "attribs": Object {
      "action": "#",
    },
    "children": Array [
      Node {
        "attribs": Object {},
        "children": Array [
          Node {
            "attribs": Object {
              "name": "title",
            },
            "children": Array [],
            "name": "textarea",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Node {
              "attribs": Object {},
              "children": Array [
                Node {
                  "data": "0/300",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "name": undefined,
            },
            "x-attribsPrefix": Object {
              "name": undefined,
            },
          },
          Node {
            "attribs": Object {},
            "children": Array [
              Node {
                "data": "0/300",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": Node {
              "attribs": Object {
                "name": "title",
              },
              "children": Array [],
              "name": "textarea",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "name": undefined,
              },
              "x-attribsPrefix": Object {
                "name": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {},
            "x-attribsPrefix": Object {},
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Node {
          "attribs": Object {},
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": Node {
            "attribs": Object {},
            "children": Array [
              Node {
                "attribs": Object {
                  "disabled": "",
                  "type": "submit",
                },
                "children": Array [
                  Node {
                    "data": "Создать тему",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "button",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "disabled": undefined,
                  "type": undefined,
                },
                "x-attribsPrefix": Object {
                  "disabled": undefined,
                  "type": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": [Circular],
            "type": "tag",
            "x-attribsNamespace": Object {},
            "x-attribsPrefix": Object {},
          },
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
      Node {
        "attribs": Object {},
        "children": Array [],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Node {
          "attribs": Object {},
          "children": Array [
            Node {
              "attribs": Object {
                "disabled": "",
                "type": "submit",
              },
              "children": Array [
                Node {
                  "data": "Создать тему",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "button",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "disabled": undefined,
                "type": undefined,
              },
              "x-attribsPrefix": Object {
                "disabled": undefined,
                "type": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
        "parent": [Circular],
        "prev": Node {
          "attribs": Object {},
          "children": Array [
            Node {
              "attribs": Object {
                "name": "title",
              },
              "children": Array [],
              "name": "textarea",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Node {
                "attribs": Object {},
                "children": Array [
                  Node {
                    "data": "0/300",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "name": undefined,
              },
              "x-attribsPrefix": Object {
                "name": undefined,
              },
            },
            Node {
              "attribs": Object {},
              "children": Array [
                Node {
                  "data": "0/300",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": Node {
                "attribs": Object {
                  "name": "title",
                },
                "children": Array [],
                "name": "textarea",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "name": undefined,
                },
                "x-attribsPrefix": Object {
                  "name": undefined,
                },
              },
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
        "type": "tag",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
      Node {
        "attribs": Object {},
        "children": Array [
          Node {
            "attribs": Object {
              "disabled": "",
              "type": "submit",
            },
            "children": Array [
              Node {
                "data": "Создать тему",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "button",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "disabled": undefined,
              "type": undefined,
            },
            "x-attribsPrefix": Object {
              "disabled": undefined,
              "type": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Node {
          "attribs": Object {},
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": Node {
            "attribs": Object {},
            "children": Array [
              Node {
                "attribs": Object {
                  "name": "title",
                },
                "children": Array [],
                "name": "textarea",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": Node {
                  "attribs": Object {},
                  "children": Array [
                    Node {
                      "data": "0/300",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "div",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": [Circular],
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "name": undefined,
                },
                "x-attribsPrefix": Object {
                  "name": undefined,
                },
              },
              Node {
                "attribs": Object {},
                "children": Array [
                  Node {
                    "data": "0/300",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": Node {
                  "attribs": Object {
                    "name": "title",
                  },
                  "children": Array [],
                  "name": "textarea",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": [Circular],
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "name": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "name": undefined,
                  },
                },
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": [Circular],
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {},
            "x-attribsPrefix": Object {},
          },
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
        "type": "tag",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
    ],
    "name": "form",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "action": undefined,
    },
    "x-attribsPrefix": Object {
      "action": undefined,
    },
  },
  "_root": LoadedCheerio {
    "0": Node {
      "children": Array [
        Node {
          "attribs": Object {},
          "children": Array [
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "head",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "body",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "body",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "head",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
          ],
          "name": "html",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
